File: LeetCode_Algorithm/136(single_number)/136.js
var singleNumber = function(nums) {
  // sol1 O(n), O(n) 76~89ms 84~68%
  // hashMap = new Map()
  // for (let num of nums){
  //   if (hashMap.has(num) == false){
  //     hashMap.set(num, 1)
  //   }else{
  //     hashMap.set(num, hashMap.get(num)+1)
  //   }
    
  // }
  // for (let key of hashMap.keys()){
  //   if (hashMap.get(key) == 1){
  //     return key
  //   }
  // }

  // sol2 O(n), O(n) 89~166ms 67~16%
  // totalSet = new Set(nums)
  // totalSet = Array.from(totalSet) // array.from => set to array
  // // totalSet = [...totalSet] // ES6 ...은 iterative한 객체를 하나씩 리턴한다.
  // // totalSetArray = new Array() // 배열로 변경한 뒤 다른 배열로 뽑아내어 복사
  // // Array.from(totalSet).forEach(function(elem, index, array){
  // //   totalSetArray.push(elem)
  // // })
  // // console.log(totalSetArray)
  // totalSetSum = totalSet.reduce((a,b)=>a+b, 0)
  // totalNumsSum = nums.reduce((a,b)=> a+b, 0)
  // return 2*totalSetSum - totalNumsSum

  // sol3 O(n), O(1) 71~104ms  89~49%
  // bit manipulation
  // num끼리 xor 연산을 한 뒤 그 수를 출력
  var ans = 0
  for (let num of nums){
    ans ^= num
    // num.toString(2) //2진수로 변환

  }
  return ans
  

}

console.log(singleNumber([1,2,4,2,1]))

File: LeetCode_Algorithm/136(single_number)/136.py
from collections import defaultdict
from typing import List

class Solution:
    def singleNumber(self, nums: List[int]) -> int:

        # sol1 O(n), O(n) 143~157ms 82~71%
        # hashmap 
        # hashMap = defaultdict(int) # 정수형 맵으로 초기화
        # for num in nums: 
        #     hashMap[num] += 1

        # for key in hashMap:
        #     if hashMap[key] == 1: return key 

        # sol2 O(n+n), O(n+n) 128~136ms 98~90%
        # set을 활용하여 *2 한 뒤 nums의 합계를 빼줌
        # return 2*sum(set(nums)) - sum(nums)

        # sol3 O(n), O(1) 163~180ms 67~58%
        # bit manipulation
        # a ^ 0 = a
        # a ^ a = 0
        # xor 연산은 같은 수끼리의 연산에서 0이 되고 
        # 0과의 연산에서는 같은 수가 나온다.
        ans = 0
        for num in nums:
            ans ^= num
        return ans


print(Solution().singleNumber([1,2,3,2,1]))

